warehouse/macaroons/auth_policy.py

```diff
@@ -0,0 +1,150 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import base64
+
+from pyramid.authentication import CallbackAuthenticationPolicy
+from pyramid.interfaces import IAuthenticationPolicy, IAuthorizationPolicy
+from pyramid.security import Denied
+from pyramid.threadlocal import get_current_request
+from zope.interface import implementer
+
+from warehouse.cache.http import add_vary_callback
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.macaroons.services import InvalidMacaroon
+
+
+def _extract_basic_macaroon(auth):
+    """
+    A helper function for extracting a macaroon from a
+    HTTP Basic Authentication-style header.
+
+    Returns ``None`` if the header doesn't contain a structurally
+    valid macaroon, or the candidate (not yet verified) macaroon.
+    """
+    try:
+        authorization = base64.b64decode(auth)
+    except ValueError:
+        return None
+
+    try:
+        auth_method, auth = authorization.split(b":", 1)
+    except ValueError:
+        return None
+
+    if auth_method != b"macaroon":
+        return None
+
+    return auth
+
+
+def extract_http_macaroon(request):
+    """
+    A helper function for the extraction of HTTP Macaroon from a given request.
+    Returns either a ``None`` if no macaroon could be found, or the byte string
+    that represents our Macaroon.
+    """
+    authorization = request.headers.get("Authorization")
+    if not authorization:
+        return None
+
+    try:
+        auth_method, auth = authorization.split(" ", 1)
+    except ValueError:
+        return None
+
+    if auth_method.lower() == "basic":
+        return _extract_basic_macaroon(auth)
```

Comment:
It looks like this function operates bytes internally but you pass unicode...

Reply:
`_extract_basic_macaroon` can take either bytes or a string, since it shoves its input into `b64decode` and the latter is happy to take either. The raw macaroon in both cases ends up being passed to `Macaroon.deserialize` (which can also take either bytes or a string), but for the sake of consistency I'll make sure that `extract_http_macaroon` only ever returns a string.